bugfix: use cmake version from dependency

trws opened the pull request:
Ensure that build tools with module-level commands in spack use the version built as part of their build graph if one exists.  This is now also required for mesa, scons, cmake and ctest, out of graph versions of these tools in path will not be found unless added as an external.  

This bug appeared because a new version of rocprim needs cmake 3.16, while I have 3.14 in my path I had added an external for cmake 3.20 to the dag, but 3.14 was still used to configure rocprim causing it to fail.  As far as I can tell, all the build tools added in build_environment.py had this problem, despite the fact that they should have been resolving these tools by name with a path search and find the one in the dag that way.  I'm still investigating why the path searching and Executable logic didn't do it, but this makes three of the build systems much more explicit, and leaves only gmake and ninja as dependencies from out in the system while ensuring the version in the dag is used if there is one.

The additional sqlite version is to perturb the hash of python to work around a relocation bug which will be fixed in a subsequent PR.

trws committed:
set module.cmake to cmake actually in the dag

tgamblin commented on var/spack/repos/builtin/packages/cmake/package.py:
```diff
@@ -375,6 +375,11 @@ def install(self, spec, prefix):
+    def setup_dependent_package(self, module, dependent_spec):
+        """Called before cmake packages's install() methods."""
+
+        module.cmake = Executable(self.spec.prefix.bin.cmake)
```
If we're doing this, would it not also make sense to just remove the cmake assignment from `build_enviroment.py`?

trws replied:
I hesitated on that lest it create fallout, but it's worth a shot.

eugeneswalker commented:
I wonder if this would fix our issue whereby `raja +rocm` fails in cmake phase because it can't find `rocprim`. I'm trying a build now, with this PR.

See comment on this line:
https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml#L225

trws committed:
remove cmake from build_environment.py

trws committed:
fix broken mock packages

eugeneswalker commented:
> I wonder if this would fix our issue wherein `raja +rocm` fails in cmake phase because it can't find `rocprim`.

Looks like this PR gets me past this error in cmake phase. I get a new error now, but that is progress!
```
==> Installing raja-2022.03.0-qeg5yfid7aadpw4fspjbfacul5hbxmpc
==> No binary for raja-2022.03.0-qeg5yfid7aadpw4fspjbfacul5hbxmpc found: installing from source
==> No patches needed for raja
==> raja: Executing phase: 'initconfig'
==> raja: Executing phase: 'cmake'
==> raja: Executing phase: 'build'
...
  >> 131    g++: error: unrecognized command-line option '--offload-arch=gfx900'
  >> 132    make[2]: *** [CMakeFiles/RAJA.dir/build.make:79: CMakeFiles/RAJA.dir/src/AlignedRangeIndexSetBuilders.cpp.o] Error 1
     133    make[2]: *** Waiting for unfinished jobs....
  >> 134    g++: error: unrecognized command-line option '--offload-arch=gfx900'
  >> 135    make[2]: *** [CMakeFiles/RAJA.dir/build.make:135: CMakeFiles/RAJA.dir/
...
```

becker33 reviewed (changes requested):
I think we need to do the same thing for `ctest` in this PR.

Doesn't need to be this PR, but looks like we need to do the same thing for `make`/`gmake`/`scons`/`ninja`/`meson`. For `make` and `gmake` we probably should keep the ones in `build_environment` but overwrite them if it's in the DAG, the others we should treat like this.

trws commented:
I agree @becker33, will take a look at those others.

trws commented:
Ah… That’s a different problem, it’s because you’re telling raja 
to build rocm with gcc.  There’s nothing in HipPackage or anything 
like that which actually changes the C++ compiler to the correct value, 
most other rocm packages are just lying about what compiler they’re 
using right now.  If you switch it to build raja with the compiler 
installed by llvm-amdgpu it will work, still working on what the best 
way around that is for the spack package and raja cmake stuff.  No clue 
why we weren’t seeing this before.

On 26 Jul 2022, at 9:01, eugeneswalker wrote:

>> I wonder if this would fix our issue wherein `raja +rocm` fails in 
>> cmake phase because it can't find `rocprim`.
>
> Looks like this PR gets me past this error in cmake phase. I get a new 
> error now, but that is progress!
> ```
> ==> Installing raja-2022.03.0-qeg5yfid7aadpw4fspjbfacul5hbxmpc
> ==> No binary for raja-2022.03.0-qeg5yfid7aadpw4fspjbfacul5hbxmpc 
> found: installing from source
> ==> No patches needed for raja
> ==> raja: Executing phase: 'initconfig'
> ==> raja: Executing phase: 'cmake'
> ==> raja: Executing phase: 'build'
> ...
>   >> 131    g++: error: unrecognized command-line option 
> '--offload-arch=gfx900'
>   >> 132    make[2]: *** [CMakeFiles/RAJA.dir/build.make:79: 
> CMakeFiles/RAJA.dir/src/AlignedRangeIndexSetBuilders.cpp.o] Error 1
>      133    make[2]: *** Waiting for unfinished jobs....
>   >> 134    g++: error: unrecognized command-line option 
> '--offload-arch=gfx900'
>   >> 135    make[2]: *** [CMakeFiles/RAJA.dir/build.make:135: 
> CMakeFiles/RAJA.dir/
> ...
> ```
>
>
> -- 
> Reply to this email directly or view it on GitHub:
> https://urldefense.us/v3/__https://github.com/spack/spack/pull/31739*issuecomment-1195674204__;Iw!!G2kpM7uM-TzIFchu!gtE2xNkgOMXgAf1bTF31xhuX1D0UTTV-PSUJC-eKNTuAN3YjLODGWcZZWQcOZUaa4Q$
> You are receiving this because you were assigned.
>
> Message ID: ***@***.***>

eugeneswalker commented:
> Ah… That’s a different problem, it’s because you’re telling raja to build rocm with gcc. There’s nothing in HipPackage or anything like that which actually changes the C++ compiler to the correct value, most other rocm packages are just lying about what compiler they’re using right now. If you switch it to build raja with the compiler installed by llvm-amdgpu it will work, still working on what the best way around that is for the spack package and raja cmake stuff. No clue why we weren’t seeing this before.
> […](#)
> On 26 Jul 2022, at 9:01, eugeneswalker wrote: > I wonder if this would fix our issue wherein `raja +rocm` fails in > cmake phase because it can't find `rocprim`. Looks like this PR gets me past this error in cmake phase. I get a new error now, but that is progress! ``` ==> Installing raja-2022.03.0-qeg5yfid7aadpw4fspjbfacul5hbxmpc ==> No binary for raja-2022.03.0-qeg5yfid7aadpw4fspjbfacul5hbxmpc found: installing from source ==> No patches needed for raja ==> raja: Executing phase: 'initconfig' ==> raja: Executing phase: 'cmake' ==> raja: Executing phase: 'build' ... >> 131 g++: error: unrecognized command-line option '--offload-arch=gfx900' >> 132 make[2]: *** [CMakeFiles/RAJA.dir/build.make:79: CMakeFiles/RAJA.dir/src/AlignedRangeIndexSetBuilders.cpp.o] Error 1 133 make[2]: *** Waiting for unfinished jobs.... >> 134 g++: error: unrecognized command-line option '--offload-arch=gfx900' >> 135 make[2]: *** [CMakeFiles/RAJA.dir/build.make:135: CMakeFiles/RAJA.dir/ ... ``` -- Reply to this email directly or view it on GitHub: [https://urldefense.us/v3/__https://github.com/spack/spack/pull/31739*issuecomment-1195674204__;Iw!!G2kpM7uM-TzIFchu!gtE2xNkgOMXgAf1bTF31xhuX1D0UTTV-PSUJC-eKNTuAN3YjLODGWcZZWQcOZUaa4Q$](https://urldefense.us/v3/__https://github.com/spack/spack/pull/31739*issuecomment-1195674204__;Iw!!G2kpM7uM-TzIFchu!gtE2xNkgOMXgAf1bTF31xhuX1D0UTTV-PSUJC-eKNTuAN3YjLODGWcZZWQcOZUaa4Q%24) You are receiving this because you were assigned. Message ID: ***@***.***>

What about doing it like this?

* https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/exago/package.py#L133-L134
* https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/hiop/package.py#L195-L196
* https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/trilinos/package.py#L465-L471

trws committed:
fix ctest, do scons, mesa and ninja as well

trws commented:
Yes, I think it would.

trws commented:
I take that back @eugeneswalker, it would *almost* work.  RAJA needs the actual clang compiler from hip rather than hipcc, so it would need a bit of extra logic to grab that depending on whether the result is `amdclang++` or `clang++`.

trws committed:
use dependency gmake/make if available

Also work around mishandling of spaces in Executable.

trws commented:
@tgamblin, @becker33 I think I've updated everything that can reasonably be done here.  The gmake package overrides the version used, but leaves the base one there so we don't break the world.  I went to add a `bmake` to the bsd make package, but it currently doesn't build for me under spack (the exact same command outside spack works, so I have no idea why) but we didn't have that before so it's not terribly important.

All of: cmake, ctest, scons, meson and ninja are removed from build_environment.py.  It would not surprise me if we find out that ninja being missing breaks something, if that happens I intend to add it as a dependency in the CmakePackage class, that class needs some cleanup anyway to get rid of all the make/ninja crud, but it would almost certainly take care of all the packages that need updates.

tgamblin commented:
@trws can you do one more thing (sorry this is currently manual) -- can you do a rebuild like we did for your `-Werror` PR?  I feel like we need that for due diligence to make sure this does not break builds.

trws commented:
All raja/camp changes have been removed from here and will be part of a subsequent PR once umpire issues are resolved.  It was proving problematic for several stacks, so this will fix the cmake/build tool issue only.

trws committed:
deal with windows

trws committed:
fix finding ninja, and probably other stuff, in windows where people like to jam things into prefix rather than prefix/bin

fix style

trws commented:
This should be ready for a re-review if anyone has time.  The failure is a syntax error in one configuration of mesa, and it's the only error that comes up both for this *and* for the build-the-world PR.  It's here: https://gitlab.spack.io/spack/spack/-/jobs/2924593

I'll look into that, but the fact that it's only AWS aarch64 where it happens and that it only happens in one configuration, and has nothing to do with the build tool selected, makes me think it's probably not directly this.

Especially look over the changes in `which`, it was necessary either to special-case `ninja` or do that, but it seems Visual Studio installs quite a number of tools that way, including others found by this PR.

trws committed:
add new sqlite version to perturb python hash

trws commented:
@becker33, if you get a chance could you take another look here? Mainly a look at the windows handling in Executable would be good.  It's finally passing builds (assuming the GHA tests don't time out this time).

becker33 commented on lib/spack/spack/util/executable.py:
```diff
@@ -291,6 +291,13 @@ def which_string(*args, **kwargs):
+        if sys.platform == "win32":
+            new_path = path[:]
+            for p in path:
+                if os.path.basename(p) == "bin":
+                    new_path.append(os.path.dirname(p))
```
Doesn't this need an `else` case? If the path doesn't end in `bin`, I don't think we should just drop it.

trws replied:
The new_path list always contains the original contents of path, hence the init with `path[:]`, then if it ends in `bin` it gets duplicated without the `bin` on the end.

becker33 commented on var/spack/repos/builtin.mock/packages/mpich/package.py:
```diff
@@ -4,6 +4,9 @@
 from spack.package import *
+from spack.util.executable import Executable
+
+ninja = Executable("ninja")  # to appease ninja test
```
Can you explain what this is doing?

trws replied:
All of the ninja build_system tests assume that a ninja `Executable` object exists in the module scope of this mock package.  Adding it as a dependency doesn't work, because it can't actually build it, so the build_system tests need us to find the ninja executable in the environment.  This was my initial stopgap for keeping them all working, but it doesn't make me particularly happy.

trws replied:
Had a bit more of a think about this and reworked the ninja handling, which took some tweaking of `which` and a couple other spots, but I think puts us in a much better place.  Apparently we weren't actually setting `-j` reliably on ninja unless it was being set by the caller, that's fixed, also put it back in the default environment if it's available because it really is a bottom dependency, especially on windows.  Bit more churn, but I like the result a whole lot more.

becker33 commented on var/spack/repos/builtin/packages/ninja/package.py:
```diff
@@ -69,3 +69,9 @@ def install(self, spec, prefix):
+        name = "ninja"
+        if sys.platform == "win32":
+            name = name + ".exe"
+        module.ninja = which(name, path=[self.spec.prefix.bin], required=True)
```
Should we make `which` just automatically try with/without `.exe` on windows?

trws replied:
That's a really good point, I'll check that and add it if we aren't doing it.

trws replied:
Apparently we are, so I'm removing this check.

trws committed:
remove extra ninja check

trws committed:
MakeExecutable and ninja refactor

This cleans up the nasty ninja module var in the mpich mock file, and
improves our usage of ninja a bit.  Specifically this is everything in
here:

* Add a new keyword argument to which to allow a custom subclass of
  Executable to be used
* Use MakeExecutable with ninja so that we set concurrency on it, which
  we apparently weren't doing before :grimace:
* Add ninja back into the default build_environment when it's available
  so that the build_system tests work, this really belongs as a bottom
  dependency anyway, *especially* on windows

trws committed:
remove no longer used import

trws committed:
refactor ninja MakeExecutable setup

trws commented:
@tgamblin, @becker33, this is now passing everything except Kokkos 3.4 on oneapi, so it's for all intents and purposes passing.  The ninja/which code has been tweaked a bit once more, but mainly to remove things.

trws commented:
Yes... I think I remember how to set that up, will copy you from a new 
PR to poke all the builders.

On 26 Jul 2022, at 13:33, Todd Gamblin wrote:

> @trws can you do one more thing (sorry this is currently manual) -- 
> can you do a rebuild like we did for your `-Werror` PR?  I feel like 
> we need that for due diligence to make sure this does not break 
> builds.
>
> -- 
> Reply to this email directly or view it on GitHub:
> https://urldefense.us/v3/__https://github.com/spack/spack/pull/31739*issuecomment-1195952364__;Iw!!G2kpM7uM-TzIFchu!k-DM0tmi6m7Re-A-p4Ez981DF4xz-1nq9KTj6_oGVqECUhgsf15TGK4E5CzKqAPiiQ$
> You are receiving this because you were mentioned.
>
> Message ID: ***@***.***>
